src/service/position-observer/position-observer-impl.js

```diff
@@ -0,0 +1,303 @@
+      this.onHostMessageHandler_(event);
+    }));
+  }
+
+  /**
+   * Callback function that gets called when unobserve last observed element.
+   * @private
+   */
+  stopCallback_() {
+    this.viewport_.disconnect();
+    while (this.unlisteners_.length) {
+      const unlisten = this.unlisteners_.pop();
+      unlisten();
+    }
+  }
+
+  /**
+   * @param {!Element} element
+   * @param {PositionObserverFidelity} fidelity
+   */
+  changeFidelity(element, fidelity) {
+    for (let i = 0; i < this.entries_.length; i++) {
+      const entry = this.entries_[i];
+      if (entry.element == element) {
+        entry.fidelity = fidelity;
+        if (fidelity == PositionObserverFidelity.HIGH) {
+          entry.turn = 0;
+        }
+        return;
+      }
+    }
+    dev().error(TAG, 'cannot change fidelity on unobserved element');
+  }
+
+  /**
+   * This should always be called in vsync.
+   * @param {boolean=} force
+   * @param {boolean=} opt_remeasure
+   * @visibleForTesting
+  */
+  updateAllEntries(force, opt_remeasure) {
+    for (let i = 0; i < this.entries_.length; i++) {
+      const entry = this.entries_[i];
+      if (opt_remeasure) {
+        entry.element.inIframePositionRect = null;
+      }
+      if (!force && entry.turn != 0) {
+        // Not ready for their turn yet.
+        entry.turn--;
+        continue;
+      }
+      // Reset entry.turn value.
+      if (!force) {
+        if (entry.fidelity == PositionObserverFidelity.LOW) {
+          entry.turn = LOW_FIDELITY_FRAME_COUNT;
+        }
+      }
+      this.updateSingleEntry_(entry);
+    }
+  }
+
+  /**
+   * To update the position of single element.
+   * Called when updateAllEntries, or when first observe an element.
+   * @param {!Object} entry
+   * @private
+   */
+  updateSingleEntry_(entry) {
+    if (this.needRefreshOnMessage_) {
+      entry.element.inIframePositionRect = null;
+    }
+
+    const elementBox =
+        this.viewport_.getLayoutRect(entry.element);
+
+    const viewportSize = this.viewport_.getSize();
+    if (!elementBox || !viewportSize) {
+      // Viewport is not ready yet.
+      return;
+    }
+    const viewportBox =
+        layoutRectLtwh(0, 0, viewportSize.width, viewportSize.height);
+
+    // Return { positionRect: <LayoutRectDef>, viewportRect: <LayoutRectDef>}
+    entry.trigger(
+      /** @type {./position-observer-entry.PositionInViewportEntryDef}*/ ({
+        positionRect: elementBox,
+        viewportRect: viewportBox,
+        relativePos: '',
+      }));
+  }
+
+  /**
+   * Handle viewport scroll event
+   * @private
+   */
+  onScrollHandler_() {
+    this.needRefreshOnMessage_ = true;
+    this.boundStopScroll_();
+    this.inScroll_ = true;
+    if (!this.measure_) {
+      this.schedulePass_();
+    }
+  }
+
+  /**
+   * Handle viewport resize event
+   * @private
+   */
+  onResizeHandler_() {
+    this.needRefreshOnMessage_ = true;
+    this.vsync_.measure(() => {
```

Comment:
But in my current code, `#schedulePass_` actively remeasure element position in every animation frame until scroll stops. I can modify the `#schedulePass_` function, let me know if you like that better, would become `#schedulePass_(isOneTime)`